src/win/device.c

```diff
@@ -0,0 +1,360 @@
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <assert.h>
+#include <stdlib.h>
+#include <fcntl.h>
+
+#include "uv.h"
+#include "internal.h"
+#include "handle-inl.h"
+#include "stream-inl.h"
+#include "req-inl.h"
+
+/* A zero-size buffer for use by uv_device_read */
+static char uv_zero_[] = "";
+
+int uv_device_open(uv_loop_t* loop,
+                   uv_device_t* device,
+                   uv_os_fd_t fd,
+                   int flags) {
+  int uvflags = 0;
+
+  assert(device);
+  if (flags != O_RDONLY && flags != O_WRONLY && flags != O_RDWR) {
+    return -EINVAL;
+  }
+
+  uv_stream_init(loop, (uv_stream_t*) device, UV_DEVICE);
+  device->handle = fd;
+  device->read_buffer.base = NULL;
+  device->read_buffer.len = 0;
+
+  if (flags & O_RDONLY) {
+    uvflags |= UV_HANDLE_READABLE;
+  } else if (flags & O_WRONLY) {
+    uvflags |= UV_HANDLE_WRITABLE;
+  } else if (flags & O_RDWR) {
+    uvflags |= UV_HANDLE_READABLE | UV_HANDLE_WRITABLE;
+  }
+
+  /* Try to associate with IOCP. */
+  if (!CreateIoCompletionPort(device->handle,
+                              loop->iocp,
+                              (ULONG_PTR) device,
+                              0)) {
+    DWORD err = GetLastError();
+    if (err) {
+      return uv_translate_sys_error(err);
+    }
+  }
+
+  device->flags |= uvflags;
+  uv_connection_init((uv_stream_t*) device);
+  return 0;
+}
+
+int uv_device_init(uv_loop_t* loop,
+                   uv_device_t* device,
+                   const char* path,
+                   int flags) {
+  HANDLE *handle;
+  DWORD dwCreationDisposition = 0;
+
+  assert(device);
+  if (flags != O_RDONLY && flags != O_WRONLY && flags != O_RDWR) {
+    return -EINVAL;
+  }
+
+  if (flags & O_RDONLY) {
+    dwCreationDisposition |= GENERIC_READ;
+  } else if (flags & O_WRONLY) {
+    dwCreationDisposition |= GENERIC_WRITE;
+  } else if (flags & O_RDWR) {
+    dwCreationDisposition |= (GENERIC_READ | GENERIC_WRITE);
+  }
+
+  handle = CreateFile(path,
+                      dwCreationDisposition,
+                      0,
+                      NULL,
+                      OPEN_EXISTING,
+                      FILE_ATTRIBUTE_SYSTEM | FILE_FLAG_OVERLAPPED,
```

Comment:
Is it needed for opening TAP devices or does it require elevated privileges?
On Oct 11, 2015 5:16 AM, "George Zhao" <notifications@github.com> wrote:

> In src/win/device.c
> <https://github.com/libuv/libuv/pull/484#discussion_r41708455>:
>
> > +  }
> > +
> > +  if (flags & O_RDONLY) {
> > +    dwCreationDisposition |= GENERIC_READ;
> > +  } else if (flags & O_WRONLY) {
> > +    dwCreationDisposition |= GENERIC_WRITE;
> > +  } else if (flags & O_RDWR) {
> > +    dwCreationDisposition |= (GENERIC_READ | GENERIC_WRITE);
> > +  }
> > +
> > +  handle = CreateFile(path,
> > +                      dwCreationDisposition,
> > +                      0,
> > +                      NULL,
> > +                      OPEN_EXISTING,
> > +                      FILE_ATTRIBUTE_SYSTEM | FILE_FLAG_OVERLAPPED,
>
> According to https://msdn.microsoft.com/en-us/library/ee332330(VS.85).aspx
> FILE_ATTRIBUTE_SYSTEM: The file is part of or used exclusively by an
> operating system.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/pull/484/files#r41708455>.
>